lib/tpm2_hmac_auth.c

```diff
@@ -0,0 +1,206 @@
+    }
+
+    return cpHash;
+}
+
+bool tpm2_read_transient_persistent_obj_name(TSS2_SYS_CONTEXT *sapi_context,
+    TPM2_HANDLE entity_handle, TPM2B_NAME *entity_name) {
+    TPM2B_PUBLIC object_public;
+    TSS2L_SYS_AUTH_RESPONSE sessions_data_out;
+    TPM2B_NAME qualified_name = TPM2B_TYPE_INIT(TPM2B_NAME, name);
+    TSS2_RC rval = TSS2_RETRY_EXP( Tss2_Sys_ReadPublic(sapi_context, entity_handle, 
+                                        NULL, &object_public, entity_name, &qualified_name, 
+                                        &sessions_data_out));
+    if (rval != TPM2_RC_SUCCESS) {
+        LOG_PERR(Tss2_Sys_ReadPublic, rval);
+        return false;
+    }
+    return true;
+}
+
+bool tpm2_get_entity_name(TSS2_SYS_CONTEXT *sapi_context, TPM2_HANDLE entity_handle, 
+    TPM2B_NAME *entity_name) {
+    TPM2B_NV_PUBLIC nv_public = TPM2B_EMPTY_INIT;
+    bool res;
+    uint8_t temp[4];
+    //Get Entity names from handles
+    switch(entity_handle & 0xFF000000) {
+        case 0x01000000: res = tpm2_util_nv_read_public(sapi_context, 
+                                                        entity_handle, 
+                                                        &nv_public,
+                                                        entity_name);
+                    if (!res) {
+                        LOG_ERR("Failed reading NV public when calculating name");
+                        return false;
+                    }
+                    break;
+        case 0x80000000:
+        case 0x81000000: res = tpm2_read_transient_persistent_obj_name(sapi_context,
+                                    entity_handle, entity_name);
+                         if (!res) {
+                            LOG_ERR("Failed reading Object public when calculating name");
+                            return false;
+                        }
+                        break;
+        case 0x00000000:
+        case 0x02000000:
+        case 0x03000000:
+        case 0x40000000: entity_name->size = sizeof(TPM2_HANDLE);
+                         memcpy(temp, (uint8_t *)&entity_handle, sizeof(TPM2_HANDLE));
+                         for(unsigned i=0; i<sizeof(TPM2_HANDLE); i++) {
+                            entity_name->name[i] = temp[3-i];
+                         }
+                         break;
+        default:         LOG_ERR("Handle not valid for HMAC auth");
+                         return false;
+    }
+    return true;
+}
+
+bool command_authentication_hmac(tpm2_session *hmac_session,
+                                 uint8_t *cpHash,
+                                 TPMS_AUTH_COMMAND *session_data) {
+
+    uint8_t *to_hmac_buffer = (unsigned char*)malloc(SHA256_DIGEST_LENGTH
+                            + tpm2_session_get_nonceTPM_size(hmac_session)
+                            + sizeof(TPMA_SESSION));
+    if (!to_hmac_buffer) {
+        LOG_ERR("oom");
+        return false;
+    }
+
+    memcpy(to_hmac_buffer, cpHash, SHA256_DIGEST_LENGTH);
```

Comment:
Looks like SHA256 is the hard coded choice. Could we make the API specify the algorithm and then just set a default in the caller.

something like this in the tool:
```
hmac_doit(a, b, c);
```
static hmac_doit(int a, int b int c, algorithm alg) {
   do stuff;
}

int hmac_doit(int a, int b int c) {
   return hmac_doit_alg(a, b, c, SHA256);
}
```

This way later if we need to, we can support HMAC in algorithm flexible way by just exposing the algorithm variant.